tsl/test/expected/gapfill.out

```diff
@@ -1080,3 +1070,338 @@ FROM metrics_tstz m1 WHERE device_id=1 GROUP BY 1,2;
+ Wed Dec 29 04:00:00 1999 PST
+ Wed Dec 29 16:00:00 1999 PST
+ Thu Dec 30 04:00:00 1999 PST
+ Thu Dec 30 16:00:00 1999 PST
+ Fri Dec 31 04:00:00 1999 PST
+ Fri Dec 31 16:00:00 1999 PST
+ Sat Jan 01 04:00:00 2000 PST
+ Sat Jan 01 16:00:00 2000 PST
+ Sun Jan 02 04:00:00 2000 PST
+ Sun Jan 02 16:00:00 2000 PST
+(11 rows)
+
+-- timestamptz time_bucket_gapfill with more complex expression
+SELECT
+  time_bucket_gapfill('12h'::interval,t)
+FROM (VALUES ('1999-12-30'::timestamptz),('2000-01-01'::timestamptz)) v(t)
+WHERE
+  t >= '2000-01-03'::timestamptz - '4d'::interval AND t < '2000-01-03'
+GROUP BY 1;
+     time_bucket_gapfill      
+------------------------------
+ Wed Dec 29 16:00:00 1999 PST
+ Thu Dec 30 04:00:00 1999 PST
+ Thu Dec 30 16:00:00 1999 PST
+ Fri Dec 31 04:00:00 1999 PST
+ Fri Dec 31 16:00:00 1999 PST
+ Sat Jan 01 04:00:00 2000 PST
+ Sat Jan 01 16:00:00 2000 PST
+ Sun Jan 02 04:00:00 2000 PST
+ Sun Jan 02 16:00:00 2000 PST
+(9 rows)
+
+-- int32 time_bucket_gapfill with greater for start
+SELECT
+  time_bucket_gapfill(1,t)
+FROM (VALUES (1),(2)) v(t)
+WHERE
+  t > -2 AND t < 3
+GROUP BY 1;
+ time_bucket_gapfill 
+---------------------
+                  -1
+                   0
+                   1
+                   2
+(4 rows)
+
+-- test actual table
+SELECT
+  time_bucket_gapfill(1,time)
+FROM metrics_int
+WHERE time >=0 AND time < 2
+GROUP BY 1;
+ time_bucket_gapfill 
+---------------------
+                   0
+                   1
+(2 rows)
+
+-- test with table alias
+SELECT
+  time_bucket_gapfill(1,time)
+FROM metrics_int m
+WHERE m.time >=0 AND m.time < 2
+GROUP BY 1;
+ time_bucket_gapfill 
+---------------------
+                   0
+                   1
+(2 rows)
+
+-- test with 2 tables
+SELECT
+  time_bucket_gapfill(1,m.time)
+FROM metrics_int m, metrics_int m2
+WHERE m.time >=0 AND m.time < 2
+GROUP BY 1;
+ time_bucket_gapfill 
+---------------------
+                   0
+                   1
+(2 rows)
+
+\set ON_ERROR_STOP 0
+-- test with 2 tables AND where clause doesnt match gapfill argument
+SELECT
+  time_bucket_gapfill(1,m2.time)
+FROM metrics_int m, metrics_int m2
+WHERE m.time >=0 AND m.time < 2
+GROUP BY 1;
+ERROR:  invalid time_bucket_gapfill argument: could not deduce start boundary from WHERE clause
+-- test inner join AND where clause doesnt match gapfill argument
+SELECT
+  time_bucket_gapfill(1,m2.time)
+FROM metrics_int m1 INNER JOIN metrics_int m2 ON m1.time=m2.time
+WHERE m1.time >=0 AND m1.time < 2
+GROUP BY 1;
+ERROR:  invalid time_bucket_gapfill argument: could not deduce start boundary from WHERE clause
+-- test inner join AND where clause doesnt match gapfill argument
+-- only toplevel where clause constraints are supported atm
+SELECT
+  time_bucket_gapfill(1,m2.time)
+FROM metrics_int m1 INNER JOIN metrics_int m2 ON m1.time=m2.time AND m2.time >=0 AND m2.time < 2
+GROUP BY 1;
+ERROR:  invalid time_bucket_gapfill argument: could not deduce start boundary from WHERE clause
+\set ON_ERROR_STOP 1
```

Comment:
I think we need some more tests:
1) What happens if the where clause has ORs instead of ANDs? (should only apply to top-level ANDed statements right.
2) What happens if we have time columns references multiple times? i..e.
 ```
SELECT
  time_bucket_gapfill(1::int8,t::int8)
FROM (VALUES (1),(2)) v(t)
WHERE
  t >= -1 AND t < 3 and t>1 AND t <=4
GROUP BY 1;
```

Reply:
yes only top-level ANDs and first one wins, i'll add tests

Reply:
tests added